File: top-interview-questions-medium/design/serialize-and-deserialize-binary-tree/380.js
/**
 * Initialize your data structure here.
 */
var RandomizedSet = function() {
  this.nums = []
  this.pos = {}
}

/**
 * Inserts a value to the set. Returns true if the set did not already contain the specified element.
 * @param {number} val
 * @return {boolean}
 */
RandomizedSet.prototype.insert = function(val) {
  if (this.pos[val] !== undefined) return false
  this.nums.push(val)
  this.pos[val] = this.nums.length - 1
  return true
}

/**
 * Removes a value from the set. Returns true if the set contained the specified element.
 * @param {number} val
 * @return {boolean}
 */
RandomizedSet.prototype.remove = function(val) {
  if (this.pos[val] === undefined) return false
  const valIdx = this.pos[val]
  const lastItem = this.nums[this.nums.length - 1]
  this.nums[valIdx] = lastItem
  this.pos[lastItem] = valIdx
  this.nums.pop()
  delete this.pos[val]
  return true
}

/**
 * Get a random element from the set.
 * @return {number}
 */
RandomizedSet.prototype.getRandom = function() {
  const idx = Math.floor(Math.random() * this.nums.length)
  return this.nums[idx]
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * var obj = Object.create(RandomizedSet).createNew()
 * var param_1 = obj.insert(val)
 * var param_2 = obj.remove(val)
 * var param_3 = obj.getRandom()
 */

File: top-interview-questions-medium/design/serialize-and-deserialize-binary-tree/380.py
import random


class RandomizedSet(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.array = []
        self.idx_dict = {}

    def insert(self, val):
        """
        Inserts a value to the set. Returns true if the set did not already contain the specified element.
        :type val: int
        :rtype: bool
        """
        if val not in self.idx_dict:
            self.array.append(val)
            self.idx_dict[val] = len(self.array) - 1
            return True

        return False

    def remove(self, val):
        """
        Removes a value from the set. Returns true if the set contained the specified element.
        :type val: int
        :rtype: bool
        """
        if val in self.idx_dict:
            idx = self.idx_dict[val]
            self.array[idx], self.array[len(self.array) - 1] = self.array[len(self.array) - 1], self.array[idx]
            self.idx_dict[self.array[idx]] = idx
            self.array.pop()
            self.idx_dict.pop(val, None)
            return True

        return False

    # Assume there are element in the array
    def getRandom(self):
        """
        Get a random element from the set.
        :rtype: int
        """
        ran_idx = random.randint(0, len(self.array) - 1)
        return self.array[ran_idx]
